bin/download_proteins_entrez.py

```diff
@@ -0,0 +1,407 @@
+    parser.add_argument("-k", "--key", required=True, help="NCBI key to allow faster access.")
+    parser.add_argument(
+        "-p",
+        "--proteins",
+        required=True,
+        metavar="FILE",
+        help="Output file (compressed) containing: protein_tmp_id, protein_sequence.",
+    )
+    parser.add_argument(
+        "-ta",
+        "--taxa_assemblies",
+        required=True,
+        metavar="FILE",
+        type=argparse.FileType("w"),
+        help="Output file containing: taxon_id, assembly_id.",
+    )
+    parser.add_argument(
+        "-ep",
+        "--entities_proteins",
+        required=True,
+        metavar="FILE",
+        type=argparse.FileType("w"),
+        help="Output file containing: protein_tmp_id, entity_name (taxon_id).",
+    )
+    parser.add_argument(
+        "-me",
+        "--microbiomes_entities",
+        required=True,
+        metavar="FILE",
+        type=argparse.FileType("w"),
+        help="Output file containing: entity_name (taxon_id), microbiome_id, entity_weight.",
+    )
+    return parser.parse_args(args)
+
+
+# get assembly length ("total_length") from entrez
+def get_assembly_length(assemblyId):
+    success = False
+    for attempt in range(3):
+        try:
+            with Entrez.efetch(db="assembly", rettype="docsum", id=assemblyId) as entrez_handle:
+                assembly_stats = Entrez.read(entrez_handle, validate=False)
+            time.sleep(1)  # avoid getting blocked by ncbi
+            success = True
+            break
+        except HTTPError as err:
+            if 500 <= err.code <= 599:
+                print("Received error from server %s" % err)
+                print("Attempt %i of 3" % attempt)
+                time.sleep(10)
+            else:
+                raise
+    if not success:
+        sys.exit("Entrez efetch download failed!")
+
+    root = ET.fromstring("<root>" + str(assembly_stats["DocumentSummarySet"]["DocumentSummary"][0]["Meta"]) + "</root>")
+    return int(root.find("./Stats/Stat[@category='total_length'][@sequence_tag='all']").text)
+
+
+def main(args=None):
+    args = parse_args(args)
+
+    # setup entrez email
+    Entrez.api_key = args.key
+    Entrez.email = args.email
+
+    # read taxonomic ids for download (together with abundances) and write 'microbiomes_entities' output
+    taxIds = []
+    input_taxids_assemblyids = {}
+    print("entity_name", "microbiome_id", "entity_weight", sep="\t", file=args.microbiomes_entities)
+    for taxid_input, microbiomeId in zip(args.taxid_input, args.microbiome_ids):
+        reader = csv.DictReader(taxid_input, delimiter="\t")
+        for row in reader:
+            # If the abundance is not defined in the taxid_input it is assigned as 1.
+            if row.keys() == set(["taxon_id"]):
+                taxIds.append(row["taxon_id"])
+                print(row["taxon_id"], microbiomeId, 1, sep="\t", file=args.microbiomes_entities, flush=True)
```

Comment:
If writing these files is a bottleneck, maybe consider an f-string and the `write` function in python

Reply:
It's currently not a bottleneck, but also wouldn't create the `write` function a memory bottleneck, as the file needs to be `open` ?

Reply:
You are right, ignore my comment 😬